File: 2015/Day01/Part2/c/task2.c
// HU: A sztenderd bemeneti/kimeneti és a standard könyvtári funkciók importálása.
// EN: Including the standard input/output and standard library headers.
#include <stdio.h>
#include <stdlib.h> // HU: Az exit() függvényhez / EN: For the exit() function

int main() {
    // HU: Egy nagy karaktertömb (C-stílusú string) a fájl tartalmának tárolására.
    // EN: A large character array (C-style string) to store the file's content.
    char c[1000000];
    
    // HU: Kezdeti változók az emelet és a pozíció követésére.
    // EN: Initial variables for tracking the floor and the position.
    int current_floor = 0;
    int pos = 0;
    
    // HU: Fájlmutató deklarálása.
    // EN: Declaring a file pointer.
    FILE *fptr;

    // HU: Megnyitjuk az input.txt fájlt olvasásra ('r' mód) és ellenőrizzük, hogy sikeres volt-e.
    // EN: We open the input.txt file for reading ('r' mode) and check if it was successful.
    if ((fptr = fopen("input.txt", "r")) == NULL) {
        printf("Error! File cannot be opened.");
        // HU: A program kilép 1-es hibakóddal, ha a fájlt nem lehet megnyitni.
        // EN: The program exits with error code 1 if the file cannot be opened.
        exit(1);
    }

    // HU: Beolvassuk a szöveget a fájlból a sortörés karakterig ('\n') a 'c' tömbbe.
    // EN: We read the text from the file into the 'c' array until a newline character ('\n') is encountered.
    fscanf(fptr, "%[^\n]", c);
    
    // HU: Egy 'for' ciklussal végigmegyünk a beolvasott 'c' string minden karakterén.
    // EN: We loop through every character of the read 'c' string with a 'for' loop.
    // HU: A ciklus addig fut, amíg el nem éri a string végét jelző '\0' karaktert.
    // EN: The loop runs until it reaches the '\0' character which marks the end of the string.
    for (int i = 0; c[i] != '\0'; i++) {
        // HU: Növeljük a pozíciót minden karakter feldolgozása előtt (1-től induló indexelés).
        // EN: We increment the position before processing each character (for 1-based indexing).
        pos++;
        
        // HU: Az 'i'-edik karaktert vizsgáljuk minden ciklusban.
        // EN: We examine the 'i'-th character in each iteration.
        if(c[i] == '(') {
            // HU: Ha a karakter '(', Santa felmegy egy emeletet.
            // EN: If the character is '(', Santa goes up one floor.
            current_floor++;
        }
        else if (c[i] == ')') {
            // HU: Ha a karakter ')', Santa lemegy egy emeletet.
            // EN: If the character is ')', Santa goes down one floor.
            current_floor--;
        }

        // HU: Minden lépés után ellenőrizzük, hogy a Mikulás a pincébe (-1. emelet) ért-e.
        // EN: After each step, we check if Santa has entered the basement (floor -1).
        if (current_floor == -1) {
            // HU: Ha igen, azonnal kilépünk a ciklusból, mert az első ilyen pozíciót kerestük.
            // EN: If yes, we break out of the loop immediately, because we were looking for the first such position.
            break;
        }
    }

    // HU: Kiírjuk a pozíciót, ahol a ciklus megszakadt.
    // EN: We print the position at which the loop was broken.
    printf("%d\n", pos);

    // HU: Bezárjuk a fájlt, hogy felszabadítsuk az erőforrást.
    // EN: We close the file to free up the resource.
    fclose(fptr);

    // HU: A program sikeresen lefutott.
    // EN: The program executed successfully.
    return 0;
}

File: 2015/Day01/Part2/python/task2.py
# HU: Kezdeti változók beállítása.
# EN: Setting up the initial variables.
current_floor = 0  # HU: Az aktuális emelet, 0-ról indul. / EN: The current floor, starts from 0.
pos = 0            # HU: A karakter pozíciójának számlálója. A feladat 1-től indexel, ezért minden lépés után növeljük.
                   # EN: Counter for the character position. The puzzle uses 1-based indexing, so we increment after each step.

try:
    # HU: A fájl biztonságos megnyitása és beolvasása.
    # EN: Safely opening and reading the file.
    with open("input.txt", 'r', encoding='utf-8') as f:
        instructions = f.read()

        # HU: Végigmegyünk az utasítások (karakterek) sorozatán.
        # EN: We loop through the sequence of instructions (characters).
        for instruction in instructions:
            
            # HU: A pozíciót minden karakter feldolgozása ELŐTT növeljük, hogy az első karakter az 1. pozíció legyen.
            # EN: We increment the position counter BEFORE processing each character, so the first character is at position 1.
            pos += 1

            # HU: Ha a karakter '(', felmegyünk egy emeletet.
            # EN: If the character is '(', we go up one floor.
            if instruction == '(':
                current_floor += 1
            # HU: Ha a karakter ')', lemegyünk egy emeletet.
            # EN: If the character is ')', we go down one floor.
            elif instruction == ')':
                current_floor -= 1

            # HU: A feladat kulcsa: minden lépés UTÁN ellenőrizzük, hogy elértük-e a -1. emeletet.
            # EN: The key to the puzzle: AFTER each step, we check if we have reached the basement (floor -1).
            if current_floor == -1:
                # HU: Ha igen, azonnal megszakítjuk a ciklust, mert csak az első ilyen alkalom érdekel minket.
                # EN: If yes, we immediately break the loop, as we are only interested in the first time this happens.
                break
    
    # HU: Kiírjuk a pozíciót, ahol a ciklus megszakadt.
    # EN: We print the position at which the loop was broken.
    print(pos)

# HU: Hibakezelés, ha a fájl nem található.
# EN: Error handling in case the file is not found.
except FileNotFoundError:
    print("Error: input.txt not found.\nHiba: input.txt nem található.")